contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+
+Before we move on to talk about Move Semantics and rvalue references, we will first briefly go through the 2 other modes available in C++, *Value Semantics* and *References Semantics*.
+
+### Value Semantics
+Value (or copy) semantics is given to users by default when we declare arugments with only the type name (e.g. `int`, `string`). It is the programming style where we are only concerned about the values stored in the objects, rather than the object itself. As such, we will always create an extra copy of the value whenever we pass it to a function (also known as [pass-by-value](http://www.learncpp.com/cpp-tutorial/72-passing-arguments-by-value/)) or when constructing a new object and variable. This ensures that each object (or function) will have their own copy value to use, without having to concern themselves with their originator.
+
+Some of the advantages of using Value Semantics is that it ensures that we no memory management issue, as we won't have any [dangling references](https://www.quora.com/What-is-dangling-reference) to objects that may not exist and no [memory leaks](https://www.geeksforgeeks.org/what-is-memory-leak-how-can-we-avoid/).
+
+Value Semantics also ensures [referential transparency](https://en.wikipedia.org/wiki/Referential_transparency), since having our own copy of the value ensures changing the value in 1 object and function will not affect the original value. This is especially useful in a [multi-threaded](https://stackoverflow.com/questions/1313062/what-is-a-multithreaded-application) environment, as this prevents removes the need for synchronization of the values, allowing the program to run faster.
+
+Another seemingly odd benefit of Value Semantics is that it may speedup programs. If we require accessing the value many times, having your own local copy of the value may be faster than having a pointer to the value and dereferencing it each time. This is especially true in C++ as it encourages Value Semantics, with optimization techniques such as [copy elision](http://en.cppreference.com/w/cpp/language/copy_elision) to help passing-by-value be faster.
+
+However, [1 major flaw](https://www.quora.com/What-are-the-drawbacks-of-pass-by-value-result) of Value Semantics and passing-by-value is having poor performance and scability. When we wish to only read the values without modifying it, creating an additional copy just for this purpose is an unnecessary computation, especially if we scale up to larger objects.
+
+### Reference Semantics
+Reference (or pointer) semantics is another mode available for users in C++. C++ allows users to declare [pointers](http://www.cplusplus.com/doc/tutorial/pointers/) and [references](https://www.geeksforgeeks.org/references-in-c/) to the address of the value. As such, we can pass around these pointers to functions and objects, and all of them will refer to and use the same value (address).
```

Comment:
>is another mode available for users

Hmm, it's more of a choice available to users rather than a mode right?

>allows users to declare pointers and references to the address of the value

This can be phrased like so:
`allows users to declare pointers and references that point to the address of a value`

and all of them will refer to and use the same ~~value~~ **address**